include/optimization/cutting_plane.h

```diff
@@ -0,0 +1,1049 @@
+
+        int i = 0;
+
+        for (; points_it != points.end(); points_it++, products_it++) {
+            if (*products_it <= b) {
+                sum_PointsInNewRegion = sum_PointsInNewRegion + points_it->getCoefficients();
+                i++;
+            }
+            else {
+                sum_PointsInCutRegion = sum_PointsInCutRegion + points_it->getCoefficients();
+            }
+        }
+
+
+        sum = sum_PointsInCutRegion + sum_PointsInNewRegion;
+        sum_PointsInNewRegion /= (double) i;
+        
+        Point point(sum_PointsInNewRegion);
+
+        return point;
+    }
+
+
+    /**
+     * Computes the quantites y = (1/N) * Sum {p | p in points}
+     * and Y = (1/N) * Sum [(p-y)* (p -y)^T], p in points
+     *
+     * These quantities will be used to set the direction vector of hit and run
+     *
+     * @tparam Point class Point
+     * @tparam NT The numeric type
+     * @param points a collection of points
+     * @param dotProducts a list with the dot products of c with all elements of points
+     * @param sum of elems in points
+     * @return sqrt(Y)
+     */
+    template<class Point, typename NT>
+    MT getIsotropicQuantities(VT &c, std::list<Point> &points, std::pair<Point, Point> &pointsPair, std::list<NT> &dotProducts) {
+        class std::list<Point>::iterator points_it = points.begin();
+        typename std::list<NT>::iterator products_it = dotProducts.begin();
+
+        int dim = points_it->dimension();
+
+        NT currentMinimum = c.dot(pointsPair.second.getCoefficients());
+        VT y;
+        y.setZero(dim);
+
+        for (auto p : points)
+            y += p.getCoefficients() / (double) points.size();
+
+        // compute Y
+        MT Y;
+        Y.setZero(dim, dim);
+        VT temp(dim);
+
+        for (auto pit = points.begin(); pit != points.end(); pit++) {
+            temp = pit->getCoefficients() - y;
+            Y = Y + (temp * temp.transpose());
+        }
+
+        Y = Y / (double) (points.size());
+
+        try {
+            Y = Y.sqrt();
+        }
+        catch (int e) {
+            throw e;
+        }
+
+        return Y;
+    }
+
+
+
+    /**
+     * Check if new point p is a better approximation than min1, min2 and if yes change min1, min2
+     *
+     *
+     * @tparam Point
+     * @tparam NT
+     * @param p
+     * @param dotProduct1 dot product of min1 with objective function
+     * @param dotProduct2 dot product of min2 with objective function
+     * @param min1 the point that minimizes the objective function
+     * @param min2 the second best point
+     * @param newProduct dot product of p with objective function
+     * @param changedMin1 set true if changed min1
+     * @param changedMin2 set true if changed min2
+     */
+    template <class Point, typename NT>
+    void getNewMinimizingPoints(const Point &p, NT& dotProduct1, NT& dotProduct2, Point &min1, Point &min2, NT newProduct, bool& changedMin1, bool& changedMin2) {
```

Comment:
`newProduct` could be a `const` reference here. Please do this in all your code wherever it is possible.